src/iris_api/sender/cache.py

```diff
@@ -346,6 +375,166 @@ def __call__(self, message, seen=None):
+                self.rates[key][1].append(0)
+
+            logging.info('Refreshed app quotas: %s', self.rates.keys())
+            sleep(60)
+
+    def __call__(self, message):
+        application = message.get('application')
+
+        if not application:
+            return True
+
+        rate = self.rates.get(application)
+
+        if not rate:
+            return True
+
+        hard_buckets, soft_buckets, hard_limit, soft_limit, wait_time, plan_name, target = rate
+
+        # Increment both buckets for this minute
+        hard_buckets[-1] += 1
+        soft_buckets[-1] += 1
+
+        # If hard limit breached, disallow sending this message and create incident
+        if sum(hard_buckets) > hard_limit:
+            self.notify_incident(application, hard_limit, len(hard_buckets), plan_name, wait_time)
+            return False
+
+        # If soft limit breached, just notify owner and still send
+        if sum(soft_buckets) > soft_limit:
+            self.notify_target(application, soft_limit, len(soft_buckets), *target)
+            return True
+
+        return True
+
+    def notify_incident(self, application, limit, duration, plan_name, wait_time):
+        logger.warning('Application %s breached hard quota. Will create incident using plan %s', application, plan_name)
+
+        # Avoid creating new incident if we have an incident that's either not claimed or claimed and wait_time hasn't been exceeded
+        last_incident = self.last_incidents.get(application)
+        if last_incident:
+            last_incident_id, last_incident_created = last_incident
+            try:
+                r = self.iris_client.get('incidents/%s' % last_incident_id)
+                r.raise_for_status()
+                last_incident = r.json()
+            except (requests.exceptions.RequestException, ValueError):
+                logger.exception('Failed looking up last created incident %s for application %s during hard quota breach', last_incident_id, application)
+            else:
+                try:
+                    if last_incident['active']:
+                        logger.info('Skipping creating incident for application %s as existing incident %s is not claimed', application, last_incident_id)
+                        return
+
+                    if wait_time and (time() - last_incident_created) < wait_time:
+                        logger.info('Skipping creating incident for application %s as it is not yet %s seconds since existing incident %s was claimed',
+                                    application, wait_time, last_incident_id)
+                        return
+                except KeyError:
+                    logger.exception('Failed parsing info for incident %s during application %s hard quota breach', last_incident_id, application)
+
+        # Make a new incident
+        try:
+            r = self.iris_client.post('incidents', json={
+                'plan': plan_name,
+                'context': {
```

Comment:
I think the templates would work out better if the metadata were more contained.

example:

{ 'quota-breach': {
    'application': application,
    'limit': limit,
    'duration': duration}
}

That way the data for a given "type" of message is contained within one node and not spread out/sharing multiple nodes.